bob-apps/bin/brt.rs

```diff
@@ -250,12 +367,216 @@ fn try_main() -> AnyResult<()> {
+            "n" | "N" => return Ok(false),
+            _ => writeln!(stdout, "Unknown command, try again")?,
+        }
+    }
+}
+
+fn move_and_recover_blob<P, Q>(
+    invalid_path: P,
+    backup_path: Q,
+    validate_every: usize,
+) -> AnyResult<()>
+where
+    P: AsRef<Path>,
+    Q: AsRef<Path>,
+{
+    std::fs::rename(&invalid_path, &backup_path).with_context(|| "move invalid blob")?;
+
+    log::info!("backup saved to {:?}", backup_path.as_ref());
+    recovery_blob(&backup_path, &invalid_path, validate_every)?;
+    Ok(())
+}
+
+fn get_backup_path(path: &str, suffix: &str) -> AnyResult<PathBuf> {
+    let mut i = 0;
+    loop {
+        let path = if i > 0 {
+            PathBuf::from_str(&format!("{}{}.{}", path, suffix, i))
+        } else {
+            PathBuf::from_str(&path)
+        }?;
+        if !path.exists() {
+            return Ok(path);
+        }
+        i += 1;
+    }
+}
+
+fn validate_blob_command(settings: &ValidateBlobSettings) -> AnyResult<()> {
+    let result = validate_recursive(&settings.path, &settings.blob_suffix, validate_blob)?;
+
+    if !result.is_empty() {
+        println!("Corrupted blobs:");
+        for path in &result {
+            println!("{}", path);
+        }
+    } else {
+        println!("All blobs is valid");
+    }
+
+    if settings.fix && !result.is_empty() {
+        if settings.confirm && !ask_confirmation("Fix invalid blob files?")? {
+            return Ok(());
+        }
+
+        let mut fails = vec![];
+        for path in &result {
+            let backup_path = get_backup_path(path, &settings.backup_suffix)?;
+            match move_and_recover_blob(&path, &backup_path, settings.validate_every) {
+                Err(err) => {
+                    log::error!("Error: {}", err);
+                    fails.push(path);
+                }
+                _ => {
+                    log::info!("[{}] recovered, backup saved to {:?}", path, backup_path);
+                }
+            }
+        }
+        if !fails.is_empty() {
+            println!("Failed to recover these blobs:");
+            for path in fails {
+                println!("{}", path);
+            }
+        }
+    }
+    Ok(())
+}
+
+fn validate_index_command(settings: &ValidateIndexSettings) -> AnyResult<()> {
+    let result = validate_recursive(&settings.path, &settings.index_suffix, validate_index)?;
+
+    if !result.is_empty() {
+        println!("Corrupted index files:");
+        for path in &result {
+            println!("{}", path);
+        }
+    } else {
+        println!("All index files is valid");
+    }
+
+    if settings.delete && !result.is_empty() {
+        if settings.confirm && !ask_confirmation("Delete invalid index files?")? {
+            return Ok(());
+        }
+
+        let mut fails = vec![];
+        for path in result {
+            match std::fs::remove_file(&path) {
+                Ok(_) => {
+                    log::info!("[{}] index file removed", path);
```

Comment:
In some places there are `log::info`, in others - `println!`, it's better to use only one method of output (preferably logs)

Reply:
Now `println` is only used to print the list of corrupted blobs or indexes, since `env_logger` prints messages to `stderr`, you can use the list of corrupted blobs in other scripts via pipe.